lib/Core/BuildEngine.cpp

```diff
@@ -904,6 +904,13 @@ class BuildEngineImpl : public BuildDBDelegate {
               ruleInfo->keyID, ruleInfo->rule, ruleInfo->result, &error);
           if (!result) {
             delegate.error(error);
+
+            // Decrement our count of outstanding tasks. This must be done prior
+            // the subsequent synchronous cancellation, since it will otherwise
+            // deadlock waiting for the task we have already popped off the
+            // queue. rdar://problem/50203529
+            --numOutstandingUnfinishedTasks;
+
```

Comment:
Doesn’t cancellation also do this? Won’t that lead to an accounting error?

Reply:
It only does so for items in the `finishedTaskInfos` vector. Since we have already removed it at this point, that is what causes the deadlock.

Reply:
Got it, thanks!